src/library/scala/collection/SeqView.scala

```diff
@@ -112,6 +114,77 @@ object SeqView {
+    outer =>
+
+    private[this] class ReverseSorted extends SeqView[A] {
+      private[this] lazy val _reversed = new SeqView.Reverse(_sorted)
+
+      def apply(i: Int): A = _reversed.apply(i)
+      def length: Int = elems.length
+      def iterator: Iterator[A] = Iterator.empty ++ _reversed.iterator // very lazy
+      override def knownSize: Int = elems.knownSize
+      override def isEmpty: Boolean = elems.isEmpty
+      override def to[C1](factory: Factory[A, C1]): C1 = _reversed.to(factory)
+      override def reverse: SeqView[A] = outer
+      override protected def reversed: Iterable[A] = outer
+
+      override def sorted[B1 >: A](implicit ord1: Ordering[B1]): SeqView[A] =
+        if (ord1 == Sorted.this.ord) outer
+        else if (ord1.isReverseOf(Sorted.this.ord)) this
+        else super.sorted(ord1)
+    }
+
+    @volatile private[this] var evaluated = false
+
+    private[this] lazy val _sorted: Seq[A] = {
+      val res = {
+        val len = underlying.length
+        if (len == 0) Nil
+        else if (len == 1) List(underlying.head)
+        else {
+          val arr = new Array[Any](len) // Array[Any] =:= Array[AnyRef]
+          underlying.copyToArray(arr)
+          java.util.Arrays.sort(arr.asInstanceOf[Array[AnyRef]], ord.asInstanceOf[Ordering[AnyRef]])
+          // casting the Array[AnyRef] to Array[A] and creating an ArraySeq from it
+          // is safe because:
+          //   - the ArraySeq is immutable, and items that are not of type A
+          //     cannot be added to it
+          //   - we know it only contains items of type A (and if this collection
+          //     contains items of another type, we'd get a CCE anyway)
+          //   - the cast doesn't actually do anything in the runtime because the
+          //     type of A is not known and Array[_] is Array[AnyRef]
+          immutable.ArraySeq.unsafeWrapArray(arr.asInstanceOf[Array[A]])
+        }
+      }
+      evaluated = true
+      underlying = null
+      res
+    }
+
+    private[this] def elems: SomeSeqOps[A] = {
+      val orig = underlying
+      if (evaluated) _sorted else orig
+    }
+
+    def apply(i: Int): A = _sorted.apply(i)
+    def length: Int = elems.length
+    def iterator: Iterator[A] = Iterator.empty ++ _sorted.iterator // very lazy
+    override def knownSize: Int = elems.knownSize
+    override def isEmpty: Boolean = elems.isEmpty
+    override def to[C1](factory: Factory[A, C1]): C1 = _sorted.to(factory)
+    override def reverse: SeqView[A] = new ReverseSorted
+    // we know `_sorted` is either tiny or has efficient random access,
+    //  so this is acceptable for `reversed`
+    override protected def reversed: Iterable[A] = new ReverseSorted
+
+    override def sorted[B1 >: A](implicit ord1: Ordering[B1]): SeqView[A] =
+      if (ord1 == this.ord) this
+      else if (ord1.isReverseOf(this.ord)) reverse
+      else super.sorted(ord1)
```

Comment:
should this be `new Sorted(elems)` instead of `super.sorted(ord1)`? (does not have to get into this PR; just idly wondering)

Reply:
Doesn't make any difference, does it? Both refer back to the original view instead of chaining `Sorted` unnecessarily.

Reply:
I believe `super.sorted` will wrap the view in a `new Sorted`, so that if you call `.sorted` three times in a row with unrelated `Ordering`s (and without evaluating any elements) you'll get `Sorted(Sorted(Sorted(originalSeq)))`; whereas `new Sorted(elems)` will wrap the original elements in that situation and you'll get `Sorted(originalSeq)`.

I don't know if it's a problem that `List(3, 2, 1).view.sorted(ord1).sorted(ord2).sorted(ord3)` never evaluates the intermediate sorts and never invokes any comparison methods on `ord1` or `ord2`. It seems to me that the contract of `SeqView#sorted` should not require a sort if it can be optimised away.